### PR TITLE
Speed up ParseRequestLine

### PR DESCRIPTION
**Pull request by benaadams**
dev
```
               Method |          Mean |     StdDev | Scaled |          RPS | Allocated |
--------------------- |-------------- |----------- |------- |------------- |---------- |
 PlaintextTechEmpower |   531.1228 ns |  5.7545 ns |   1.00 | 1,882,803.68 |     112 B |
           LiveAspNet |   917.0677 ns |  9.0826 ns |   1.73 | 1,090,432.07 |     112 B |
              Unicode | 1,300.7374 ns | 18.2239 ns |   2.45 |   768,794.70 |     112 B |
```
pr
```
               Method |          Mean |     StdDev | Scaled |          RPS | Allocated |
--------------------- |-------------- |----------- |------- |------------- |---------- |
 PlaintextTechEmpower |   513.2275 ns |  6.5310 ns |   1.00 | 1,948,453.72 |     112 B |
           LiveAspNet |   909.4809 ns | 10.0160 ns |   1.77 | 1,099,528.33 |     112 B |
              Unicode | 1,283.9817 ns | 13.0742 ns |   2.50 |   778,827.32 |     112 B |
```

### CODE REVIEW

**Comment by benaadams** [comment]
pr
```
               Method |          Mean |     StdDev | Scaled |          RPS | Allocated |
--------------------- |-------------- |----------- |------- |------------- |---------- |
 PlaintextTechEmpower |   503.9617 ns |  6.1930 ns |   1.00 | 1,984,277.87 |     112 B |
           LiveAspNet |   901.9177 ns |  9.6750 ns |   1.79 | 1,108,748.57 |     112 B |
              Unicode | 1,258.5205 ns | 14.2006 ns |   2.50 |   794,583.77 |     112 B |
```

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs [hunk 1]
```diff
@@ -34,219 +34,173 @@ public KestrelHttpParser(IKestrelTrace log)
...
             Span<byte> customMethod;
             var i = 0;
             var length = span.Length;
-            var done = false;
 
             fixed (byte* data = &span.DangerousGetPinnableReference())
             {
-                switch (StartLineState.KnownMethod)
+                // Method
+                if (span.GetKnownMethod(out method, out var methodLength))
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs]: Would it be faster if we use pointer in GetKnownMethod and version?

Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs]: I was just talking to @ahsonkhan about this

Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs]: Have a version version; is a bit ugly - but will push

Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs]: Pointer version of version https://github.com/aspnet/KestrelHttpServer/pull/1463/commits/75a24af18aeeab45539cee85353c9f6ddf8ed789

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/HttpUtilities.cs [hunk 2]
```diff
@@ -221,6 +221,44 @@ public static bool GetKnownVersion(this Span<byte> span, out HttpVersion knownVe
...
+        /// <remarks>
+        /// A "known HTTP version" Is is either HTTP/1.0 or HTTP/1.1.
+        /// Since those fit in 8 bytes, they can be optimally looked up by reading those bytes as a long. Once
+        /// in that format, it can be checked against the known versions.
+        /// The Known versions will be checked with the required '\r'.
+        /// To optimize performance the HTTP/1.1 will be checked first.
+        /// </remarks>
+        /// <returns><c>true</c> if the input matches a known string, <c>false</c> otherwise.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public unsafe static HttpVersion GetKnownVersion(byte* location)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/HttpUtilities.cs]: Pass in the length?

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs (hunk 1 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs]: Same for the method.

**Comment by benaadams** [comment]
Rebased

**Comment by benaadams** [comment]
dev
```
                        Method |      Mean |    StdDev | Scaled |        RPS | Allocated |
------------------------------ |---------- |---------- |------- |----------- |---------- |
          PlaintextTechEmpower | 1.4231 us | 0.0063 us |   1.00 | 702,705.85 |     353 B |
 PipelinedPlaintextTechEmpower | 1.1439 us | 0.0076 us |   0.80 | 874,220.46 |     351 B |
                    LiveAspNet | 3.0643 us | 0.0362 us |   2.15 | 326,334.31 |   1.12 kB |
           PipelinedLiveAspNet | 2.7153 us | 0.0256 us |   1.91 | 368,284.39 |   1.11 kB |
                       Unicode | 5.2271 us | 0.0370 us |   3.67 | 191,309.20 |   1.99 kB |
              UnicodePipelined | 5.0083 us | 0.1999 us |   3.52 | 199,669.47 |   1.98 kB |
```
pr
```
                        Method |      Mean |    StdDev | Scaled |        RPS | Allocated |
------------------------------ |---------- |---------- |------- |----------- |---------- |
          PlaintextTechEmpower | 1.3217 us | 0.0070 us |   1.00 | 756,626.99 |     349 B |
 PipelinedPlaintextTechEmpower | 1.0927 us | 0.0088 us |   0.83 | 915,173.46 |     348 B |
                    LiveAspNet | 2.8567 us | 0.0300 us |   2.16 | 350,049.41 |   1.11 kB |
           PipelinedLiveAspNet | 2.5057 us | 0.0140 us |   1.90 | 399,089.12 |    1.1 kB |
                       Unicode | 4.8917 us | 0.0261 us |   3.70 | 204,427.38 |   1.95 kB |
              UnicodePipelined | 4.7156 us | 0.0321 us |   3.57 | 212,062.52 |   1.97 kB |
```

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs [hunk 3]
```diff
@@ -34,219 +34,142 @@ public KestrelHttpParser(IKestrelTrace log)
...
-            handler.OnStartLine(method, httpVersion, targetBuffer, pathBuffer, query, customMethod);
+                examined = consumed;
+                return true;
+            }
+        }
 
-            consumed = end;
-            examined = consumed;
-            return true;
+        private static bool GetMultiSpan(ref ReadableBuffer buffer, ref Span<byte> span, out ReadCursor end)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs]: TryGetSpan

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs [hunk 4]
```diff
@@ -34,219 +34,142 @@ public KestrelHttpParser(IKestrelTrace log)
...
+        {
+            if (ReadCursorOperations.Seek(buffer.Start, buffer.End, out end, ByteLF) >= 0)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs]: Accessing buffer.Start twice vs storing it in a local? That intentional I assume.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs [hunk 5]
```diff
@@ -1216,15 +1216,14 @@ protected void ReportApplicationError(Exception ex)
             Log.ApplicationError(ConnectionId, ex);
         }
 
-        public void OnStartLine(HttpMethod method, HttpVersion version, Span<byte> target, Span<byte> path, Span<byte> query, Span<byte> customMethod)
+        public void OnStartLine(HttpMethod method, HttpVersion version, bool pathEncoded, Span<byte> target, Span<byte> path, Span<byte> query, Span<byte> customMethod)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Move the pathEncoded bool to last?

Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Moved

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs (hunk 4 above)
Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs]: No 😄

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/HttpUtilities.cs [hunk 6]
```diff
@@ -147,34 +147,46 @@ public static string GetAsciiStringEscaped(this Span<byte> span, int maxChars)
         /// </remarks>
         /// <returns><c>true</c> if the input matches a known string, <c>false</c> otherwise.</returns>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static bool GetKnownMethod(this Span<byte> span, out HttpMethod method, out int length)
+        public static unsafe bool GetKnownMethod(this Span<byte> span, out HttpMethod method, out int length)
         {
-            if (span.TryRead<uint>(out var possiblyGet))
+            fixed(byte* data = &span.DangerousGetPinnableReference())
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/HttpUtilities.cs]: nit:space

Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/HttpUtilities.cs]: Done

**Comment by davidfowl** [comment]
👏

**Pull request merged**